#Data Wrangling

##Gathering Data

In [55]:
import pandas as pd

In [56]:
covid_df = pd.read_csv("https://raw.githubusercontent.com/hrizriz/mini-project/main/dataset/covid_19_indonesia_time_series_all.csv")
covid_df.head()

Date Location ISO Code     Location  New Cases  New Deaths  \
0  3/1/2020             ID-JK  DKI Jakarta          2           0   
1  3/2/2020             ID-JK  DKI Jakarta          2           0   
2  3/2/2020               IDN    Indonesia          2           0   
3  3/2/2020             ID-RI         Riau          1           0   
4  3/3/2020             ID-JK  DKI Jakarta          2           0   

   New Recovered  New Active Cases  Total Cases  Total Deaths  \
0              0                 2           39            20   
1              0                 2           41            20   
2              0                 2            2             0   
3              0                 1            1             0   
4              0                 2           43            20   

   Total Recovered  ...  Latitude New Cases per Million  \
0               75  ... -6.204699                  0.18   
1               75  ... -6.204699                  0.18   
2                0  ... -0.789275                  0.01   
3                1  ...  0.511648                  0.16   
4               75  ... -6.204699                  0.18   

   Total Cases per Million New Deaths per Million Total Deaths per Million  \
0                     3.60                    0.0                     1.84   
1                     3.78                    0.0                     1.84   
2                     0.01                    0.0                     0.00   
3                     0.16                    0.0                     0.00   
4                     3.96                    0.0                     1.84   

  Total Deaths per 100rb Case Fatality Rate Case Recovered Rate  \
0                   0.18             51.28%             192.31%   
1                   0.18             48.78%             182.93%   
2                   0.00              0.00%               0.00%   
3                   0.00              0.00%             100.00%   
4                   0.18             46.51%             174.42%   

  Growth Factor of New Cases  Growth Factor of New Deaths  
0                        NaN                          NaN  
1                        1.0                          1.0  
2                        NaN                          NaN  
3                        NaN                          NaN  
4                        1.0                          1.0  

[5 rows x 38 columns]

##Assessing Data

In [57]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31822 entries, 0 to 31821
Data columns (total 38 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Date                         31822 non-null  object 
 1   Location ISO Code            31822 non-null  object 
 2   Location                     31822 non-null  object 
 3   New Cases                    31822 non-null  int64  
 4   New Deaths                   31822 non-null  int64  
 5   New Recovered                31822 non-null  int64  
 6   New Active Cases             31822 non-null  int64  
 7   Total Cases                  31822 non-null  int64  
 8   Total Deaths                 31822 non-null  int64  
 9   Total Recovered              31822 non-null  int64  
 10  Total Active Cases           31822 non-null  int64  
 11  Location Level               31822 non-null  object 
 12  City or Regency              0 non-null      float64
 13  Province        

In [58]:
covid_df.describe()

New Cases    New Deaths  New Recovered  New Active Cases  \
count  31822.000000  31822.000000   31822.000000      31822.000000   
mean     402.311388      9.920652     390.398498          1.992238   
std     2320.629838     64.139080    2199.878802       1219.513355   
min        0.000000      0.000000       0.000000     -29938.000000   
25%        3.000000      0.000000       2.000000        -12.000000   
50%       27.000000      0.000000      20.000000          0.000000   
75%      130.000000      3.000000     123.000000         19.000000   
max    64718.000000   2069.000000   61361.000000      39165.000000   

        Total Cases   Total Deaths  Total Recovered  Total Active Cases  \
count  3.182200e+04   31822.000000     3.182200e+04        31822.000000   
mean   1.594500e+05    4564.753221     1.492615e+05         5623.782415   
std    6.264435e+05   17693.731369     5.958536e+05        28537.412305   
min    1.000000e+00       0.000000     0.000000e+00        -2343.000000   
25%    5.223250e+03     128.000000     3.913500e+03           80.000000   
50%    2.359650e+04     565.500000     2.102750e+04          557.000000   
75%    6.992775e+04    2189.000000     6.414200e+04         2279.000000   
max    6.405044e+06  157876.000000     6.218708e+06       586113.000000   

       City or Regency  Total Regencies  ...  Population Density  \
count              0.0     31822.000000  ...        31822.000000   
mean               NaN        24.027277  ...          738.894928   
std                NaN        68.359734  ...         2729.431626   
min                NaN         1.000000  ...            8.590000   
25%                NaN         7.000000  ...           47.790000   
50%                NaN        11.000000  ...          103.840000   
75%                NaN        18.000000  ...          262.700000   
max                NaN       416.000000  ...        16334.310000   

          Longitude      Latitude  New Cases per Million  \
count  31822.000000  31822.000000           31822.000000   
mean     113.700478     -2.725681              28.133292   
std        9.862068      3.608065              74.309710   
min       96.910522     -8.682205               0.000000   
25%      106.109004     -6.204699               0.830000   
50%      113.417654     -2.461746               5.710000   
75%      121.201093      0.212037              22.290000   
max      138.696030      4.225615            1459.040000   

       Total Cases per Million  New Deaths per Million  \
count             31822.000000            31822.000000   
mean              11485.038800                0.640308   
std               16477.385479                1.933016   
min                   0.010000                0.000000   
25%                1291.367500                0.000000   
50%                6804.285000                0.000000   
75%               14557.360000                0.540000   
max              130231.620000               63.800000   

       Total Deaths per Million  Total Deaths per 100rb  \
count              31822.000000            31822.000000   
mean                 289.633640               28.963329   
std                  363.428724               36.342881   
min                    0.000000                0.000000   
25%                   38.862500                3.890000   
50%                  158.415000               15.840000   
75%                  389.910000               38.990000   
max                 1632.600000              163.260000   

       Growth Factor of New Cases  Growth Factor of New Deaths  
count                29883.000000                 28375.000000  
mean                     1.326795                     1.033834  
std                      2.679379                     1.351755  
min                      0.000000                     0.000000  
25%                      0.650000                     0.750000  
50%                      1.000000                     1.000000  
75%                      1.310000      

In [59]:
covid_df.isna().sum()

Date                               0
Location ISO Code                  0
Location                           0
New Cases                          0
New Deaths                         0
New Recovered                      0
New Active Cases                   0
Total Cases                        0
Total Deaths                       0
Total Recovered                    0
Total Active Cases                 0
Location Level                     0
City or Regency                31822
Province                         929
Country                            0
Continent                          0
Island                           929
Time Zone                        929
Special Status                 27264
Total Regencies                    0
Total Cities                     901
Total Districts                    0
Total Urban Villages             904
Total Rural Villages             929
Area (km2)                         0
Population                         0
Population Density                 0
L

In [60]:
print("Jumlah duplikasi: ", covid_df.duplicated().sum())

Jumlah duplikasi:  0


Mengubah kolom menjadi snake_case format

In [61]:
covid_df.columns = covid_df.columns.str.lower().str.replace(' ', '_')

Mengecek kembali apakah semua nama kolom sudah sesuai dengan yang diinginkan yaitu snake_case format

In [62]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31822 entries, 0 to 31821
Data columns (total 38 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   date                         31822 non-null  object 
 1   location_iso_code            31822 non-null  object 
 2   location                     31822 non-null  object 
 3   new_cases                    31822 non-null  int64  
 4   new_deaths                   31822 non-null  int64  
 5   new_recovered                31822 non-null  int64  
 6   new_active_cases             31822 non-null  int64  
 7   total_cases                  31822 non-null  int64  
 8   total_deaths                 31822 non-null  int64  
 9   total_recovered              31822 non-null  int64  
 10  total_active_cases           31822 non-null  int64  
 11  location_level               31822 non-null  object 
 12  city_or_regency              0 non-null      float64
 13  province        

Mengganti kolom area_(km2) menjadi area_km2

In [63]:
covid_df = covid_df.rename(columns={'area_(km2)': 'area_km2'})

Menghapus kolom 'city_or_regency' dikarenakan kolom ini tidak memiliki baris sama sekali

In [64]:
covid_df = covid_df.drop(columns=['city_or_regency','special_status'])

Mengecek kembali tabel covid_df apakah kolom 'city_or_regency' sudah terhapus dan kolom 'area_(km2) sudah berganti nama jadi 'area_km2'

In [65]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31822 entries, 0 to 31821
Data columns (total 36 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   date                         31822 non-null  object 
 1   location_iso_code            31822 non-null  object 
 2   location                     31822 non-null  object 
 3   new_cases                    31822 non-null  int64  
 4   new_deaths                   31822 non-null  int64  
 5   new_recovered                31822 non-null  int64  
 6   new_active_cases             31822 non-null  int64  
 7   total_cases                  31822 non-null  int64  
 8   total_deaths                 31822 non-null  int64  
 9   total_recovered              31822 non-null  int64  
 10  total_active_cases           31822 non-null  int64  
 11  location_level               31822 non-null  object 
 12  province                     30893 non-null  object 
 13  country         

In [66]:
covid_df.isna().sum()

date                              0
location_iso_code                 0
location                          0
new_cases                         0
new_deaths                        0
new_recovered                     0
new_active_cases                  0
total_cases                       0
total_deaths                      0
total_recovered                   0
total_active_cases                0
location_level                    0
province                        929
country                           0
continent                         0
island                          929
time_zone                       929
total_regencies                   0
total_cities                    901
total_districts                   0
total_urban_villages            904
total_rural_villages            929
area_km2                          0
population                        0
population_density                0
longitude                         0
latitude                          0
new_cases_per_million       

In [67]:
select_df = covid_df[['province','island','time_zone','total_cities', 'total_urban_villages','total_rural_villages','growth_factor_of_new_cases','growth_factor_of_new_deaths']]
select_df.head(100)

province         island  time_zone  total_cities  \
0         DKI Jakarta           Jawa  UTC+07:00           5.0   
1         DKI Jakarta           Jawa  UTC+07:00           5.0   
2                 NaN            NaN        NaN          98.0   
3                Riau       Sumatera  UTC+07:00           2.0   
4         DKI Jakarta           Jawa  UTC+07:00           5.0   
..                ...            ...        ...           ...   
95               Riau       Sumatera  UTC+07:00           2.0   
96  Sulawesi Tenggara       Sulawesi  UTC+08:00           2.0   
97               Bali  Nusa Tenggara  UTC+08:00           1.0   
98             Banten           Jawa  UTC+07:00           4.0   
99        DKI Jakarta           Jawa  UTC+07:00           5.0   

    total_urban_villages  total_rural_villages  growth_factor_of_new_cases  \
0                  267.0                   NaN                         NaN   
1                  267.0                   NaN                         1.0   
2                 8488.0               74953.0                         NaN   
3                  268.0                1591.0                         NaN   
4                  267.0                   NaN                         1.0   
..                   ...                   ...                         ...   
95                 268.0                1591.0                         1.0   
96                 377.0                1911.0                         NaN   
97                  80.0                 636.0                         1.0   
98                 313.0                1238.0                         0.0   
99                 267.0                   NaN                         1.0   

    growth_factor_of_new_deaths  
0                           NaN  
1                          1.00  
2                           NaN  
3                           NaN  
4                          1.00  
..                          ...  
95                         1.00  
96                         1.00  
97                         1.00  
98                          NaN  
99                         0.33  

[100 rows x 8 columns]

Menghapus value 98 pada total_cities dikarenakan value tersebut anomali

In [68]:
covid_df = covid_df[covid_df['total_cities'] != 98.0]

In [69]:
select_df = covid_df[['province','island','time_zone','total_cities', 'total_urban_villages','total_rural_villages','growth_factor_of_new_cases','growth_factor_of_new_deaths']]
select_df.head(100)

province      island  time_zone  total_cities  \
0                   DKI Jakarta        Jawa  UTC+07:00           5.0   
1                   DKI Jakarta        Jawa  UTC+07:00           5.0   
3                          Riau    Sumatera  UTC+07:00           2.0   
4                   DKI Jakarta        Jawa  UTC+07:00           5.0   
6                    Jawa Barat        Jawa  UTC+07:00           9.0   
..                          ...         ...        ...           ...   
111  Daerah Istimewa Yogyakarta        Jawa  UTC+07:00           1.0   
113                  Jawa Barat        Jawa  UTC+07:00           9.0   
114                 Jawa Tengah        Jawa  UTC+07:00           6.0   
115                  Jawa Timur        Jawa  UTC+07:00           9.0   
116            Kalimantan Timur  Kalimantan  UTC+08:00           3.0   

     total_urban_villages  total_rural_villages  growth_factor_of_new_cases  \
0                   267.0                   NaN                         NaN   
1                   267.0                   NaN                        1.00   
3                   268.0                1591.0                         NaN   
4                   267.0                   NaN                        1.00   
6                   645.0                5312.0                         NaN   
..                    ...                   ...                         ...   
111                  46.0                 392.0                        1.00   
113                 645.0                5312.0                        0.36   
114                 753.0                7809.0                        3.00   
115                 777.0                7724.0                         NaN   
116                 197.0                 841.0                        0.00   

     growth_factor_of_new_deaths  
0                            NaN  
1                            1.0  
3                            NaN  
4                            1.0  
6                            NaN  
..                           ...  
111                          1.0  
113                          NaN  
114                          NaN  
115                          NaN  
116                          1.0  

[100 rows x 8 columns]

In [70]:
covid_df.isna().sum()

date                              0
location_iso_code                 0
location                          0
new_cases                         0
new_deaths                        0
new_recovered                     0
new_active_cases                  0
total_cases                       0
total_deaths                      0
total_recovered                   0
total_active_cases                0
location_level                    0
province                          0
country                           0
continent                         0
island                            0
time_zone                         0
total_regencies                   0
total_cities                    901
total_districts                   0
total_urban_villages            904
total_rural_villages            929
area_km2                          0
population                        0
population_density                0
longitude                         0
latitude                          0
new_cases_per_million       

In [71]:
covid_df.total_cities.value_counts()

1.0    9931
2.0    9088
4.0    2735
9.0    1839
5.0    1833
3.0    1827
6.0     922
8.0     913
7.0     904
Name: total_cities, dtype: int64

In [72]:
covid_df[covid_df.total_cities.isna()] # cek missing value

date location_iso_code        location  new_cases  new_deaths  \
309    3/28/2020             ID-SR  Sulawesi Barat          1           0   
338    3/29/2020             ID-SR  Sulawesi Barat          0           0   
369    3/30/2020             ID-SR  Sulawesi Barat          0           0   
402    3/31/2020             ID-SR  Sulawesi Barat          0           0   
435     4/1/2020             ID-SR  Sulawesi Barat          0           0   
...          ...               ...             ...        ...         ...   
31640  9/10/2022             ID-SR  Sulawesi Barat          1           0   
31675  9/11/2022             ID-SR  Sulawesi Barat          0           0   
31710  9/12/2022             ID-SR  Sulawesi Barat          0           0   
31745  9/13/2022             ID-SR  Sulawesi Barat          0           0   
31780  9/14/2022             ID-SR  Sulawesi Barat          1           0   

       new_recovered  new_active_cases  total_cases  total_deaths  \
309                0                 1            1             0   
338                0                 0            1             0   
369                0                 0            1             0   
402                0                 0            1             0   
435                0                 0            1             0   
...              ...               ...          ...           ...   
31640              0                 1        15600           394   
31675              1                -1        15600           394   
31710              0                 0        15600           394   
31745              0                 0        15600           394   
31780              0                 1        15601           394   

       total_recovered  ...  latitude new_cases_per_million  \
309                  0  ... -2.461746                  0.64   
338                  0  ... -2.461746                  0.00   
369                  0  ... -2.461746                  0.00   
402                  0  ... -2.461746                  0.00   
435                  0  ... -2.461746                  0.00   
...                ...  ...       ...                   ...   
31640            15200  ... -2.461746                  0.64   
31675            15201  ... -2.461746                  0.00   
31710            15201  ... -2.461746                  0.00   
31745            15201  ... -2.461746                  0.00   
31780            15201  ... -2.461746                  0.64   

      total_cases_per_million new_deaths_per_million total_deaths_per_million  \
309                      0.64                    0.0                     0.00   
338                      0.64                    0.0                     0.00   
369                      0.64                    0.0                     0.00   
402                      0.64                    0.0                     0.00   
435                      0.64                    0.0                     0.00   
...                       ...                    ...                      ...   
31640                10000.10                    0.0                   252.57   
31675                10000.10                    0.0                   252.57   
31710                10000.10                    0.0                   252.57   
31745                10000.10                    0.0                   252.57   
31780                10000.74                    0.0                   252.57   

      total_deaths_per_100rb case_fatality_rate  case_recovered_rate  \
309                     0.00              0.00%                0.00%   
338                     0.00              0.00%                0.00%   
369                     0.00              0.00%                0.00%   
402                     0.00              0.00%                0.00%   
435                     0.00              0.00%                0.00%   
...                      ...                ...                  ...   
31640                  2

In [73]:
select_df = covid_df[['location','total_cities', 'total_regencies']]
select_df.head(100)

location  total_cities  total_regencies
0                   DKI Jakarta           5.0                1
1                   DKI Jakarta           5.0                1
3                          Riau           2.0               10
4                   DKI Jakarta           5.0                1
6                    Jawa Barat           9.0               18
..                          ...           ...              ...
111  Daerah Istimewa Yogyakarta           1.0                4
113                  Jawa Barat           9.0               18
114                 Jawa Tengah           6.0               29
115                  Jawa Timur           9.0               29
116            Kalimantan Timur           3.0                7

[100 rows x 3 columns]

In [74]:
covid_df['total_cities'].fillna(value=0, inplace=True) # mengganti column missing value pada total_cites menjadi 0 dikarenakan tidak ada kota madya pada provinsi tersebut
covid_df[covid_df.total_cities.isna()] # cek column missing value

Empty DataFrame
Columns: [date, location_iso_code, location, new_cases, new_deaths, new_recovered, new_active_cases, total_cases, total_deaths, total_recovered, total_active_cases, location_level, province, country, continent, island, time_zone, total_regencies, total_cities, total_districts, total_urban_villages, total_rural_villages, area_km2, population, population_density, longitude, latitude, new_cases_per_million, total_cases_per_million, new_deaths_per_million, total_deaths_per_million, total_deaths_per_100rb, case_fatality_rate, case_recovered_rate, growth_factor_of_new_cases, growth_factor_of_new_deaths]
Index: []

[0 rows x 36 columns]

In [75]:
covid_df.isna().sum()

date                              0
location_iso_code                 0
location                          0
new_cases                         0
new_deaths                        0
new_recovered                     0
new_active_cases                  0
total_cases                       0
total_deaths                      0
total_recovered                   0
total_active_cases                0
location_level                    0
province                          0
country                           0
continent                         0
island                            0
time_zone                         0
total_regencies                   0
total_cities                      0
total_districts                   0
total_urban_villages            904
total_rural_villages            929
area_km2                          0
population                        0
population_density                0
longitude                         0
latitude                          0
new_cases_per_million       

In [76]:
covid_df.total_urban_villages.value_counts()

35.0     1810
267.0     929
268.0     928
645.0     927
313.0     924
753.0     922
377.0     920
80.0      919
197.0     916
46.0      914
693.0     913
777.0     912
142.0     911
792.0     911
163.0     910
110.0     908
387.0     907
118.0     907
139.0     904
205.0     904
175.0     904
332.0     904
230.0     904
95.0      903
73.0      901
99.0      901
144.0     900
82.0      900
172.0     899
145.0     899
327.0     890
72.0      888
Name: total_urban_villages, dtype: int64

Untuk mengisi kolom 'total_urban_villages' yang hilang, akan diterapkan metode mean untuk pengisian valuenya

In [77]:
mean_value = covid_df['total_urban_villages'].mean()
covid_df['total_urban_villages'] = covid_df['total_urban_villages'].fillna(mean_value)

In [78]:
covid_df.isna().sum()

date                              0
location_iso_code                 0
location                          0
new_cases                         0
new_deaths                        0
new_recovered                     0
new_active_cases                  0
total_cases                       0
total_deaths                      0
total_recovered                   0
total_active_cases                0
location_level                    0
province                          0
country                           0
continent                         0
island                            0
time_zone                         0
total_regencies                   0
total_cities                      0
total_districts                   0
total_urban_villages              0
total_rural_villages            929
area_km2                          0
population                        0
population_density                0
longitude                         0
latitude                          0
new_cases_per_million       

Begitu juga untuk pengisian value kosong pada kolom 'total_rural_villages' akan dilakukan metode mean

In [79]:
mean_value = covid_df['total_rural_villages'].mean()
covid_df['total_rural_villages'] = covid_df['total_rural_villages'].fillna(mean_value)

In [80]:
covid_df.isna().sum()

date                              0
location_iso_code                 0
location                          0
new_cases                         0
new_deaths                        0
new_recovered                     0
new_active_cases                  0
total_cases                       0
total_deaths                      0
total_recovered                   0
total_active_cases                0
location_level                    0
province                          0
country                           0
continent                         0
island                            0
time_zone                         0
total_regencies                   0
total_cities                      0
total_districts                   0
total_urban_villages              0
total_rural_villages              0
area_km2                          0
population                        0
population_density                0
longitude                         0
latitude                          0
new_cases_per_million       

In [81]:
covid_df.growth_factor_of_new_cases.describe()

count    28958.000000
mean         1.336134
std          2.720726
min          0.000000
25%          0.640000
50%          1.000000
75%          1.320000
max        175.000000
Name: growth_factor_of_new_cases, dtype: float64

In [82]:
new_case_diatas10 = covid_df[covid_df['growth_factor_of_new_cases'] > 10]['growth_factor_of_new_cases']
new_case_diatas10

241      57.00
348      11.00
601      14.50
818      16.00
920      15.00
         ...  
31256    17.00
31750    10.50
31757    76.75
31758    12.00
31806    10.50
Name: growth_factor_of_new_cases, Length: 226, dtype: float64

Untuk pengisian value pada kolom 'growth_factor_of_new_cases' digunakan metode median

In [83]:
median_value = covid_df['growth_factor_of_new_cases'].median()
covid_df['growth_factor_of_new_cases'] = covid_df['growth_factor_of_new_cases'].fillna(median_value)

In [84]:
covid_df.isna().sum()

date                              0
location_iso_code                 0
location                          0
new_cases                         0
new_deaths                        0
new_recovered                     0
new_active_cases                  0
total_cases                       0
total_deaths                      0
total_recovered                   0
total_active_cases                0
location_level                    0
province                          0
country                           0
continent                         0
island                            0
time_zone                         0
total_regencies                   0
total_cities                      0
total_districts                   0
total_urban_villages              0
total_rural_villages              0
area_km2                          0
population                        0
population_density                0
longitude                         0
latitude                          0
new_cases_per_million       

In [85]:
covid_df.growth_factor_of_new_deaths.describe()

count    27451.000000
mean         1.031211
std          1.367264
min          0.000000
25%          0.750000
50%          1.000000
75%          1.000000
max        134.500000
Name: growth_factor_of_new_deaths, dtype: float64

In [86]:
new_death_diatas10 = covid_df[covid_df['growth_factor_of_new_deaths'] > 10]['growth_factor_of_new_deaths']
new_death_diatas10

2748      11.00
3553      10.67
4813      15.00
5203      23.00
5227      14.00
5448      15.00
5722      13.00
5833      27.00
6554      13.00
7744     134.50
8417      10.50
8447      18.00
9748      23.00
10063     23.00
10194     11.00
10688     13.00
10868     25.20
11300     11.50
11405     13.00
12337     16.75
12338     42.50
12457     15.00
12827     25.00
13002     14.57
13073     37.00
13177     10.43
13275     16.09
13283     15.33
14587     18.00
14683     10.50
14761     12.00
14902     11.50
15601     17.00
15881     23.00
16091     12.50
16157     11.00
16470     12.00
16810     10.50
16897     20.00
16962     11.00
17067     11.00
17258     12.00
17282     15.94
17371     11.00
17441     11.00
17759     14.00
17906     11.00
18106     11.00
18351     25.00
18691     11.00
18707     18.00
23777     11.00
24615     12.00
24877     22.00
25296     11.00
25926     12.00
Name: growth_factor_of_new_deaths, dtype: float64

Nilai '134.50' pada kolom 'growth_factor_of_new_deaths' sepertinya ini value anomali, jadi akan diubah menjadi 13.45

In [87]:
covid_df.growth_factor_of_new_deaths.replace(covid_df.growth_factor_of_new_deaths.max(), 13.45, inplace=True)

In [88]:
covid_df.growth_factor_of_new_deaths.describe()

count    27451.000000
mean         1.026801
std          1.107272
min          0.000000
25%          0.750000
50%          1.000000
75%          1.000000
max         42.500000
Name: growth_factor_of_new_deaths, dtype: float64

Untuk pengisian value pada kolom 'growth_factor_of_new_deaths' digunakan metode median

In [89]:
median_value = covid_df['growth_factor_of_new_deaths'].median()
covid_df['growth_factor_of_new_deaths'] = covid_df['growth_factor_of_new_deaths'].fillna(median_value)

In [90]:
covid_df.isna().sum()

date                           0
location_iso_code              0
location                       0
new_cases                      0
new_deaths                     0
new_recovered                  0
new_active_cases               0
total_cases                    0
total_deaths                   0
total_recovered                0
total_active_cases             0
location_level                 0
province                       0
country                        0
continent                      0
island                         0
time_zone                      0
total_regencies                0
total_cities                   0
total_districts                0
total_urban_villages           0
total_rural_villages           0
area_km2                       0
population                     0
population_density             0
longitude                      0
latitude                       0
new_cases_per_million          0
total_cases_per_million        0
new_deaths_per_million         0
total_deat

In [91]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30893 entries, 0 to 31820
Data columns (total 36 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   date                         30893 non-null  object 
 1   location_iso_code            30893 non-null  object 
 2   location                     30893 non-null  object 
 3   new_cases                    30893 non-null  int64  
 4   new_deaths                   30893 non-null  int64  
 5   new_recovered                30893 non-null  int64  
 6   new_active_cases             30893 non-null  int64  
 7   total_cases                  30893 non-null  int64  
 8   total_deaths                 30893 non-null  int64  
 9   total_recovered              30893 non-null  int64  
 10  total_active_cases           30893 non-null  int64  
 11  location_level               30893 non-null  object 
 12  province                     30893 non-null  object 
 13  country         

Ubah tipe data

In [92]:
covid_df['date'] = pd.to_datetime(covid_df['date'])
covid_df['location_iso_code'] = covid_df['location_iso_code'].astype(str)
covid_df['location'] = covid_df['location'].astype(str)
covid_df['location_level'] = covid_df['location_level'].astype(str)
covid_df['province'] = covid_df['province'].astype(str)
covid_df['country'] = covid_df['country'].astype(str)
covid_df['continent'] = covid_df['continent'].astype(str)
covid_df['island'] = covid_df['island'].astype(str)


covid_df['case_fatality_rate'] = covid_df['case_fatality_rate'].str.replace('%', '').astype(float) / 100
covid_df['case_recovered_rate'] = covid_df['case_recovered_rate'].str.replace('%', '').astype(float) / 100

In [97]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30893 entries, 0 to 31820
Data columns (total 36 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   date                         30893 non-null  datetime64[ns]
 1   location_iso_code            30893 non-null  object        
 2   location                     30893 non-null  object        
 3   new_cases                    30893 non-null  int64         
 4   new_deaths                   30893 non-null  int64         
 5   new_recovered                30893 non-null  int64         
 6   new_active_cases             30893 non-null  int64         
 7   total_cases                  30893 non-null  int64         
 8   total_deaths                 30893 non-null  int64         
 9   total_recovered              30893 non-null  int64         
 10  total_active_cases           30893 non-null  int64         
 11  location_level               30893 non-nu

Export data

In [98]:
covid_df.to_csv("covid_data.csv", index=False)